File: proyecto/backend/routes/activities.py
from flask import Blueprint, request, jsonify
from models import Activity, Classroom
from database import SessionLocal

bp = Blueprint("activities", __name__)

@bp.route("", methods=["GET"])
def get_activities():
    period_id = request.args.get("period_id", type=int)
    session = SessionLocal()
    try:
        query = session.query(Activity)
        if period_id:
            query = query.join(Classroom).filter(Classroom.period_id == period_id)
        activities = query.all()
        return jsonify([
            {
                "id": a.id,
                "title": a.title,
                "description": a.description,
                "due_date": a.due_date.isoformat(),
                "max_score": a.max_score,
                "classroom_id": a.classroom_id
            } for a in activities
        ])
    finally:
        session.close()

@bp.route("/", methods=["POST"])
def create_activity():
    data = request.json
    if not data.get("title") or not data.get("max_score") or not data.get("due_date"):
        return jsonify({"error": "Missing required fields"}), 400

    session = SessionLocal()
    try:
        activity = Activity(
            title=data["title"],
            description=data.get("description"),
            due_date=data["due_date"],
            max_score=data["max_score"],
            classroom_id=data.get("classroom_id")
        )
        session.add(activity)
        session.commit()
        session.refresh(activity)
        return jsonify({"id": activity.id, "message": "Activity created"}), 201
    finally:
        session.close()

# Update y Delete mantienen la lógica anterior


# ------------------ FUNCIONES CRUD NUEVAS ------------------

@bp.route("/<int:id>/", methods=["PUT"])
def update_activity(id):
    data = request.json
    session = SessionLocal()
    try:
        activity = session.query(Activity).get(id)
        if not activity:
            return jsonify({"error": "Activity not found"}), 404

        activity.title = data.get("title", activity.title)
        activity.description = data.get("description", activity.description)
        activity.due_date = data.get("due_date", activity.due_date)
        activity.max_score = data.get("max_score", activity.max_score)
        activity.classroom_id = data.get("classroom_id", activity.classroom_id)

        session.commit()
        session.refresh(activity)
        return jsonify({"id": activity.id, "message": "Activity updated"})
    finally:
        session.close()

@bp.route("/<int:id>/", methods=["DELETE"])
def delete_activity(id):
    session = SessionLocal()
    try:
        activity = session.query(Activity).get(id)
        if not activity:
            return jsonify({"error": "Activity not found"}), 404

        session.delete(activity)
        session.commit()
        return jsonify({"message": "Activity deleted"})
    finally:
        session.close()

File: proyecto/api/routes/activities.js
const { Router } = require("express");
const { db } = require("../../lib/firebase");
const { snapshotToArray, asyncHandler } = require("../middleware/auth");

const router = Router();

router.get("/", asyncHandler(async (req, res) => {
  const { period_id, classroom_id } = req.query;
  const snap = await db.ref("activities").orderByChild("teacher_id").equalTo(req.teacherId).once("value");
  let activities = snapshotToArray(snap);

  if (classroom_id) {
    activities = activities.filter(a => a.classroom_id === classroom_id);
  }

  if (period_id) {
    const clsSnap = await db.ref("classrooms").orderByChild("period_id").equalTo(period_id).once("value");
    const clsIds = new Set(snapshotToArray(clsSnap).map(c => c.id));
    activities = activities.filter(a => clsIds.has(a.classroom_id));
  }

  res.json(activities);
}));

router.post("/", asyncHandler(async (req, res) => {
  const { title, description, due_date, max_score, classroom_id } = req.body;
  if (!title || max_score === undefined || !due_date) {
    return res.status(400).json({ error: "Missing required fields" });
  }
  const ref = db.ref("activities").push();
  await ref.set({
    title,
    description: description || "",
    due_date,
    max_score: parseFloat(max_score),
    classroom_id: classroom_id || "",
    teacher_id: req.teacherId,
  });
  res.status(201).json({ id: ref.key, message: "Activity created" });
}));

router.put("/:id", asyncHandler(async (req, res) => {
  const snap = await db.ref(`activities/${req.params.id}`).once("value");
  if (!snap.val()) return res.status(404).json({ error: "Activity not found" });
  if (snap.val().teacher_id !== req.teacherId) return res.status(403).json({ error: "Forbidden" });

  const updates = {};
  if (req.body.title !== undefined) updates.title = req.body.title;
  if (req.body.description !== undefined) updates.description = req.body.description;
  if (req.body.due_date !== undefined) updates.due_date = req.body.due_date;
  if (req.body.max_score !== undefined) updates.max_score = parseFloat(req.body.max_score);
  if (req.body.classroom_id !== undefined) updates.classroom_id = req.body.classroom_id;
  await db.ref(`activities/${req.params.id}`).update(updates);
  res.json({ id: req.params.id, message: "Activity updated" });
}));

router.delete("/:id", asyncHandler(async (req, res) => {
  const snap = await db.ref(`activities/${req.params.id}`).once("value");
  if (!snap.val()) return res.status(404).json({ error: "Activity not found" });
  if (snap.val().teacher_id !== req.teacherId) return res.status(403).json({ error: "Forbidden" });
  await db.ref(`activities/${req.params.id}`).remove();
  res.json({ message: "Activity deleted" });
}));

module.exports = router;
